### PR TITLE
[pytorch][codegen] migrate gen_variable_factories.py to the new data model

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47818 [pytorch][codegen] migrate gen_variable_factories.py to the new data model**
* #47747 [pytorch][codegen] add fully migrated scripts to mypy strict config
* #47746 [pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly
* #47745 [pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model

This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24909478](https://our.internmc.facebook.com/intern/diff/D24909478)

### CODE REVIEW

File: tools/autograd/gen_variable_factories.py [hunk 1]
```diff
@@ -3,77 +3,79 @@
...
             # it is ignored anyways (and we actually have an assertion that it isn't set
             # which would fail otherwise). We handle requires_grad explicitly here
             # instead of passing it through to the kernel.
-            actual = "at::TensorOptions({}).requires_grad(c10::nullopt)".format(actual)
-        actuals.append(actual)
+            exprs.append(f'at::TensorOptions({arg.name}).requires_grad(c10::nullopt)')
+        else:
+            exprs.append(arg.name)
+
     requires_grad = "options.requires_grad()" if has_tensor_options else "false"
```
[the remark this answers is not in the document]
Comment (bhosmer) [reply tools/autograd/gen_variable_factories.py:69]: ...also, does the use of `options.requires_grad()` here contradict the comment above about ignoring it? Both are in old code, I know... just wondering if there's an easy clarification to be had. But NBD if not

**Review by bhosmer — APPROVED** [review]
Looks good! Handful of minor thoughts but nothing huge - only one non-style thing (the hardcoded "options" arg name).

File: tools/autograd/gen_variable_factories.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_variable_factories.py:69]: > ...also, does the use of `options.requires_grad()` here contradict the comment above about ignoring it? Both are in old code, I know... just wondering if there's an easy clarification to be had. But NBD if not

The comment above is talking about the `options` argument being passed into aten, which is ignored. The use of `options.requires_grad()` here is to set the requires grad bit on the result tensor. See the example:
```
inline at::Tensor empty(at::IntArrayRef size, const at::TensorOptions & options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::Tensor tensor = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::empty(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format);
  })();
  at::Tensor result =
    autograd::make_variable(std::move(tensor), /*requires_grad=*/options.requires_grad());
  return result;
}
```

File: tools/autograd/gen_variable_factories.py [hunk 2]
```diff
@@ -3,77 +3,79 @@
...
-    qualified_type = "{}at::{}".format(argument_type[:index], argument_type[index:])
-    return maybe_optional_type(qualified_type, opt_match)
+    qualified_type = f'{argument_type[:index]}at::{argument_type[index:]}'
+    return maybe_optional_type(qualified_type, opt_match is not None)
+
+def gen_variable_factories(out: str, native_yaml_path: str, template_path: str) -> None:
+    native_functions = parse_native_yaml(native_yaml_path)
+    fm = FileManager(install_dir=out, template_dir=template_path, dry_run=False)
+    fm.write_with_template('variable_factories.h', 'variable_factories.h', lambda: {
+        'generated_comment': f'@generated from {fm.template_dir}/variable_factories.h',
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_variable_factories.py:37]: > just curious why the override is necessary here... what would `write_with_template()` be plugging in for `generated_comment` by default?

The default comment is saying the file is generated by `aten/src/ATen/gen.py`

File: tools/autograd/gen_variable_factories.py [hunk 3]
```diff
@@ -3,77 +3,79 @@
...
-                    decl,
-                    has_tensor_options,
-                )
-            )
-    write(out,
-          "variable_factories.h",
-          CodeTemplate.from_file(template_path + "/variable_factories.h"),
-          {"function_definitions": function_definitions})
+    if Variant.function not in f.variants or \
+            not has_tensor_options and not name.endswith("_like"):
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_variable_factories.py:47]: done

File: tools/autograd/gen_variable_factories.py [hunk 4]
```diff
@@ -3,77 +3,79 @@
...
-    def maybe_optional_type(t, opt_match):
-        return 'c10::optional<{}>'.format(t) if opt_match else t
+# Add 'at::' to types defined in ATen namespace, e.g. Tensor, TensorList, IntArrayRef and etc.
+# TODO: maybe update the cpp argument API to take optional namespace argument?
+def fully_qualified_type(argument_type: str) -> str:
+    def maybe_optional_type(type: str, opt: bool) -> str:
+        return f'c10::optional<{type}>' if opt else type
 
     opt_match = OPTIONAL_TYPE_PATTERN.match(argument_type)
     if opt_match:
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_variable_factories.py:24]: done

File: tools/autograd/gen_variable_factories.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply tools/autograd/gen_variable_factories.py:38]: ah got it... lol seems like that defaulting could be improved but obvs not for this PR

File: tools/autograd/gen_variable_factories.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply tools/autograd/gen_variable_factories.py]: thanks! your explanation actually improves on the comment :) but not worth further pre-landing churn on your part

[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_variable_factories.py]: > thanks! your explanation actually improves on the comment :) but not worth further pre-landing churn on your part

Added some comment - having both handled in `isinstance(arg.argument, TensorOptionsArguments)` branch it should be more clear what we are trying to do.

**Pull request closed without merging**